pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -36,6 +36,64 @@ using Pennylane::LightningKokkos::Util::vector2view;
 /// @endcond
 
 namespace Pennylane::LightningKokkos::Functors {
+template <class PrecisionT, class FuncT> class applyNCNFunctor {
+    using KokkosComplexVector = Kokkos::View<Kokkos::complex<PrecisionT> *>;
+    using KokkosIntVector = Kokkos::View<std::size_t *>;
+    using ScratchViewComplex =
+        Kokkos::View<Kokkos::complex<PrecisionT> *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using ScratchViewSizeT =
+        Kokkos::View<std::size_t *,
+                     Kokkos::DefaultExecutionSpace::scratch_memory_space,
+                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using MemberType = Kokkos::TeamPolicy<>::member_type;
+
+    Kokkos::View<Kokkos::complex<PrecisionT> *> arr;
+    const FuncT core_function;
+    KokkosIntVector indices;
+    KokkosIntVector parity;
+    KokkosIntVector rev_wires;
+    KokkosIntVector rev_wire_shifts;
+    std::size_t dim;
+
+  public:
+    template <class ExecutionSpace>
+    applyNCNFunctor([[maybe_unused]] ExecutionSpace exec,
+                    Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+                    std::size_t num_qubits,
+                    const std::vector<std::size_t> &controlled_wires,
+                    const std::vector<bool> &controlled_values,
+                    const std::vector<std::size_t> &wires, FuncT core_function_)
+        : arr(arr_), core_function(core_function_) {
+
+        std::size_t two2N =
+            std::exp2(num_qubits - wires.size() - controlled_wires.size());
+        dim = std::exp2(wires.size());
+        const auto &[parity_, rev_wires_] =
+            reverseWires(num_qubits, wires, controlled_wires);
+        parity = parity_;
+        std::vector<std::size_t> indices_ =
+            generateBitPatterns(wires, num_qubits);
+        ControlBitPatterns(indices_, num_qubits, controlled_wires,
+                           controlled_values);
+        indices = vector2view(indices_);
+        std::size_t scratch_size = ScratchViewComplex::shmem_size(dim) +
```

Comment:
Is `shmem_size` used to assign the size of shared memory of GPU? If yes, how can we make it safer as shared memory size varies from GPUs to GPUs?

Reply:
When the scratch memory is set to level 0 (in L85), it is using the GPU Shared memory in the case for Nvidia GPUs. Assuming O(10KB) shared memory, this should be fine for at least 9 wires. If it does exceed this, we could in theory set the scratch memory level to 1, which is larger but slower.

The scratch_size here could actually be smaller, I will update this line from 
```
std::size_t scratch_size = ScratchViewComplex::shmem_size(dim) +  ScratchViewSizeT::shmem_size(dim);
```
to
```
std::size_t scratch_size = ScratchViewSizeT::shmem_size(dim);
```
(This kernel does not need scratch memory for the matrix like for qubitunitary)


Some further [reference](https://www.nersc.gov/assets/Uploads/KokkosTutorial_AdvancedScratchSpace.pdf) (p.8):
```Scratch memory (scratch pad) as manual cache:
▶ Accessing data in (level 0) scratch memory is (usually) much faster than global
memory.
▶ GPUs have separate, dedicated, small, low-latency scratch memories (NOT
subject to coalescing requirements).
▶ CPUs don’t have special hardware, but programming with scratch memory results
in cache-aware memory access patterns.
▶ Roughly, it’s like a user-managed L1 cache
```

Reply:
That's say, would a MultiRZ gate targets at 20 wires with only 1 control wire break the simulation?

Reply:
It might actually break it, that's a good point. So in this case I might not use scratch or I will use level 1 scratch (depending on what the performance and memory limit is)